src/accessory/defined/camera/manager.rs

```diff
@@ -0,0 +1,374 @@
+use std::{collections::HashMap, error::Error, net::IpAddr, sync::Arc};
+
+use futures::{future::BoxFuture, lock::Mutex};
+use serde_json::Value;
+use thiserror::Error;
+use uuid::Uuid;
+
+use crate::{
+    characteristic::{AsyncCharacteristicCallbacks, HapCharacteristic},
+    pointer,
+    server::Server,
+    tlv, HapType,
+};
+
+use super::{media::MediaProvider, protocol, CameraAccessory};
+
+use tokio::sync::mpsc::{channel, Sender};
+
+pub struct StreamManagerBuilder<MP>
+where
+    MP: MediaProvider,
+{
+    accessory: CameraAccessory,
+    provider: MP,
+}
+
+#[derive(Debug, Clone)]
+pub struct SessionState {
+    pub session_id: Uuid,
+    pub address: IpAddr,
+    pub video_port: u16,
+    pub audio_port: u16,
+    pub video_crypto: protocol::SetupEndpointCrypto,
+    pub audio_crypto: protocol::SetupEndpointCrypto,
+    pub video_ssrc: u32,
+    pub audio_ssrc: u32,
+}
+
+// todo: consider types to be more restricted
+#[derive(Clone, Default, Debug)]
+pub struct StreamOptions {
+    pub video: Value,
+    pub audio: Value,
+    pub srtp: bool,
+}
+
+pub struct StreamManager<MP>
+where
+    MP: MediaProvider,
+{
+    accessory: pointer::Accessory,
+    inner: Arc<Mutex<StreamManagerInner<MP>>>,
+}
+
+#[derive(Debug)]
+enum StreamingStatus {
+    Streaming,
+    Available,
+}
+
+pub struct StreamManagerInner<MP>
+where
+    MP: MediaProvider,
+{
+    streaming_status: HashMap<usize, StreamingStatus>,
+    sessions: HashMap<Uuid, SessionState>,
+    /// Setup endpoint change sender
+    sec_sender: Sender<SetupEndpointChange>,
+    /// Media provider
+    provider: MP,
+}
+
+#[derive(Debug, Error)]
+pub enum StreamManagerError {
+    #[error("Invalid accessory")]
+    InvalidAccessory,
+    #[error("Missing characteristic")]
+    MissingCharacteristic,
+    #[error("Missing session")]
+    SessionNotFound,
+    #[error("Provider error: {0}")]
+    Provider(Box<dyn Error + Send + Sync + 'static>),
+    #[error(transparent)]
+    Hap(#[from] crate::Error),
+    #[error(transparent)]
+    ProtocolError(#[from] protocol::ProtocolError),
+}
+
+struct SetupEndpointChange(usize, Vec<u8>);
+
+/// Builder adds callbacks to Camera Accessory and creates
+/// channel which allows modification of setup_endpoint characteristic
+/// from inside a callback
+impl<MP> StreamManagerBuilder<MP>
+where
+    MP: MediaProvider + Send + Sync + 'static,
+{
+    pub fn new(accessory: CameraAccessory, provider: MP) -> Self {
+        Self { accessory, provider }
+    }
+
+    /// Builder is needed because in case of Camera Accessory we have to self modify value of
+    /// setup_endpoints to create proper handshake with iOS device
+    /// Current accessory contract doesnt allow to self modify Characteristic and locks
+    /// are created when adding accessory to the server
+    pub async fn build<S>(self, server: S) -> Result<StreamManager<MP>, StreamManagerError>
```

Comment:
Hmm. How does one start do `server.run_handle()` after passing the server into the builder? I did a bit of debugging and was able to start it by using a reference to the server (`server: &S`) and `stream_builder.build(&server)` but it's unclear what the intended use is.